src/main/java/com/example/hop_oasis/service/data/RecommendationsServiceImpl.java

```diff
@@ -0,0 +1,185 @@
+    }
+
+    private Recommendations getRecommendationsForCider(Long ciderId) {
+        final var cider = ciderRepository.findById(ciderId)
+                .orElseThrow(() -> new ResourceNotFoundException("Cider not found, id: %s", ciderId));
+
+        // find bundles with the same cider
+        final var bundlesWithTheSameCider =
+                getFirstN(productBundleRepository.getBundlesWithSimilarName(cider.getCiderName()), 5);
+
+        // some snacks
+        final var randomSnacks = snackRepository.findRandomRecords(5);
+
+        // other cider or beer
+        final var randomCider = ciderRepository.findRandomRecords(2);
+        randomCider.removeIf(c -> Objects.equals(c.getId(), ciderId));
+
+        final var randomBeers = beerRepository.findRandomRecords(2);
+
+        return new Recommendations(
+                randomBeers,
+                randomCider,
+                randomSnacks,
+                bundlesWithTheSameCider);
+    }
+
+    private Recommendations getRecommendationsForSnack(Long snackId) {
+        final var snack = snackRepository.findById(snackId)
+                .orElseThrow(() -> new ResourceNotFoundException("Cider not found, id: %s", snackId));
+
+        // find bundles with the same snacks
+        final var bundlesWithTheSameSnacks =
+                getFirstN(productBundleRepository.getBundlesWithSimilarName(snack.getSnackName()), 5);
+
+        // find some beer
+        final var randomBeers = beerRepository.findRandomRecords(2);
+
+        // find random cider
+        final var randomCider = ciderRepository.findRandomRecords(2);
+
+        // find some snacks
+        final var randomSnacks = snackRepository.findRandomRecords(5);
+        randomSnacks.removeIf(c -> Objects.equals(c.getId(), snackId));
+
+        return new Recommendations(
+                randomBeers,
+                randomCider,
+                randomSnacks,
+                bundlesWithTheSameSnacks);
+    }
+
+    private Recommendations getRecommendationsForBundle(Long bundleId) {
+        final var bundle = productBundleRepository.findById(bundleId)
+                .orElseThrow(() -> new ResourceNotFoundException("Bundle not found, id: %s", bundleId));
+
+        // find bundles with the same name
+        final var similarBundles = getFirstN(productBundleRepository.getBundlesWithSimilarName(bundle.getName()), 5);
+        similarBundles.removeIf(b -> Objects.equals(b.getId(), bundleId));
+
+        // find some beer
+        final var randomBeers = beerRepository.findRandomRecords(5);
+
+        // find some cider
+        final var randomCider = ciderRepository.findRandomRecords(5);
+
+        // find some snacks
+        final var randomSnacks = snackRepository.findRandomRecords(5);
+
+        return new Recommendations(
+                randomBeers,
+                randomCider,
+                randomSnacks,
+                similarBundles);
+    }
+
+    private <T> List<T> getFirstN(List<T> list, int n) {
```

Comment:
This method is redundant since you can do all limitation at the DB level

Reply:
Unfortunately jpql doesn't support limitation , only native query does

Reply:
yeah, that's a well-known shame. You can pass Pageable as a second parameter.
You can create it with `PageRequest.of(...)`